default/python/AI/pathfinding.py

```diff
@@ -0,0 +1,134 @@
+     While searching for a path, keep track of the fleet's fuel. Compared to standard A*/dijkstra,
+     nodes are locked only for a certain minimum level of fuel. If a longer path yields a higher fuel
+     level at a given system, then that path considered for pathfinding and added to the queue.
+
+
+    :param start: start system id
+    :type start: int
+    :param target:  target system id
+    :type target: int
+    :param fleet_id: fleet to find the path for
+    :type fleet_id: int
+    :param minimum_fuel_at_target: optional - if specified, only accept paths that leave the
+                                   fleet with at least this much fuel left at the target system
+    :type minimum_fuel_at_target: int
+    :return: shortest possible path including resupply-deroutes in the form of system ids
+             including both start and target system
+    :rtype: tuple[int]
+    """
+    path_information = namedtuple('path_information', ['distance', 'fuel', 'path'])
+    universe = fo.getUniverse()
+    empire_id = fo.empireID()
+    fleet = universe.getFleet(fleet_id)
+    supplied_systems = set(fo.getEmpire().fleetSupplyableSystemIDs)
+
+    # make sure the target is connected to the start system
+    shortest_possible_path_distance = universe.shortestPathDistance(start, target)
+    if shortest_possible_path_distance == -1:
+        warn("Requested path between disconnected systems, doing nothing.")
+        return None
+
+    # make sure the minimum fuel at target is realistic
+    if minimum_fuel_at_target < 0:
+        error("Requested negative fuel at target.")
+        return None
+
+    if fleet.maxFuel < minimum_fuel_at_target:
+        return None
+
+    start_fuel = fleet.maxFuel if start in supplied_systems else fleet.fuel
+
+    # We have 1 free jump from supplied system into unsupplied systems.
+    # Thus, the target system must be at most maxFuel + 1 jumps away
+    # in order to reach the system under standard conditions.
+    # In some edge cases, we may have more supply here than what the
+    # supply graph suggests. For example, we could have recently refueled
+    # in a system that is now blockaded by an enemy or we have found
+    # the refuel special.
+    target_distance_from_supply = -min(state.get_system_supply(target), 0)
+    if (fleet.maxFuel + 1 < (target_distance_from_supply + minimum_fuel_at_target) and
+            universe.jumpDistance(start, target) > (start_fuel - minimum_fuel_at_target)):
+        # can't possibly reach this system with the required fuel
+        return None
+
+    # initialize data structures
+    path_cache = {}
+    queue = []
+
+    # add starting system to queue
+    heappush(queue, (shortest_possible_path_distance, path_information(distance=0, fuel=start_fuel, path=(start,)),
+                     start))
+
+    while queue:
+        # get next system u with path information
+        (_, path_info, current) = heappop(queue)
```

Comment:
If there is no reason to ever work with empty paths here (and I don't see any), then the final item in ```path_info.path``` will always be the same as ```current``` and I would like to suggest we keep the heap structure simpler and then here say
```python
(_, path_info) = heappop(queue)
current = path_info.path[-1]
```

Although it adds an extra line right here, in several other places (wherever we construct new entries for the queue) it keeps things simpler.   And to me it seems a bit helpful that the start of the loop has this explicit reminder of the relationship between ```path_info.path``` and ```current```.

Reply:
changed.